executor/pipelined_window.go

```diff
@@ -0,0 +1,462 @@
+		return p.curRowIdx + 1, nil
+	}
+}
+
+// produce produces rows and append it to chk, return produced means number of rows appended into chunk, available means
+// number of rows processed but not fetched
+func (p *processor) produce(ctx sessionctx.Context, chk *chunk.Chunk, remained int) (produced int64, err error) {
+	var (
+		start uint64
+		end   uint64
+	)
+	for remained > 0 && p.moreToProduce(ctx) {
+		start, err = p.getStart(ctx)
+		if err != nil {
+			return
+		}
+		end, err = p.getEnd(ctx)
+		if err != nil {
+			return
+		}
+		if end > p.rowCnt {
+			if !p.whole {
+				return
+			}
+			end = p.rowCnt
+		}
+		if start >= p.rowCnt {
+			if !p.whole {
+				return
+			}
+			start = p.rowCnt
+		}
+		// if start >= end, we should return a default value, and we reset the frame to empty.
+		if start >= end {
+			for i, wf := range p.windowFuncs {
+				if !p.emptyFrame {
+					wf.ResetPartialResult(p.partialResults[i])
+				}
+				err = wf.AppendFinalResult2Chunk(ctx, p.partialResults[i], chk)
+				if err != nil {
+					return
+				}
+			}
+			if !p.emptyFrame {
+				p.emptyFrame = true
+				p.initializedSlidingWindow = false
+			}
+		} else {
+			p.emptyFrame = false
+			for i, wf := range p.windowFuncs {
+				slidingWindowAggFunc := p.slidingWindowFuncs[i]
+				if p.lastStartRow != start || p.lastEndRow != end {
+					if slidingWindowAggFunc != nil && p.initializedSlidingWindow {
+						err = slidingWindowAggFunc.Slide(ctx, p.getRow, p.lastStartRow, p.lastEndRow, start-p.lastStartRow, end-p.lastEndRow, p.partialResults[i])
+					} else {
+						// For MinMaxSlidingWindowAggFuncs, it needs the absolute value of each start of window, to compare
+						// whether elements inside deque are out of current window.
+						if minMaxSlidingWindowAggFunc, ok := wf.(aggfuncs.MaxMinSlidingWindowAggFunc); ok {
+							// Store start inside MaxMinSlidingWindowAggFunc.windowInfo
+							minMaxSlidingWindowAggFunc.SetWindowStart(start)
+						}
+						// TODO(zhifeng): track memory usage here
+						wf.ResetPartialResult(p.partialResults[i])
+						_, err = wf.UpdatePartialResult(ctx, p.getRows(start, end), p.partialResults[i])
+					}
+				}
+				if err != nil {
+					return
+				}
+				err = wf.AppendFinalResult2Chunk(ctx, p.partialResults[i], chk)
+				if err != nil {
+					return
+				}
+			}
+			p.initializedSlidingWindow = true
+		}
+		p.curRowIdx++
+		p.lastStartRow, p.lastEndRow = start, end
+
+		produced++
+		remained--
+	}
+	extend := p.curRowIdx
+	if p.lastEndRow < extend {
+		extend = p.lastEndRow
+	}
+	if p.lastStartRow < extend {
+		extend = p.lastStartRow
+	}
+	if extend > p.rowStart {
+		p.rows = p.rows[extend-p.rowStart:]
+		p.rowStart = extend
+	}
+	return
+}
+
+func (p *processor) moreToProduce(ctx sessionctx.Context) bool {
+	if p.curRowIdx >= p.rowCnt {
+		return false
+	}
+	if p.whole {
+		return true
+	}
+	start, err := p.getStart(ctx)
+	_ = err
+	end, err := p.getEnd(ctx)
+	_ = err
+	// TODO(Zhifeng): handle the error here
```

Comment:
I don't have a good idea how to handle it (this function should just return a bool), if it returns two variable, it would make the caller dirty to look at.

Reply:
Do you hae any suggestions?

Reply:
I don't think it's dirty. For the three places which call it, checking error is necessary.